proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+      for field in $fields
+        s = align_to(s, align(field))
+        s += size(direction, field)
+      align_to(s, align(record $fields))
+    variant $cases =>
+      s = 0
+      discrim = variant_discriminant($cases)
+      for case in $cases
+        if case:
+          s = max(s, size(direction, record([discim, case])))
+        else:
+          s = max(s, size(direction, discrim))
+      s
+    handle => size(i32)
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        size(record([i32, i32, i32]))
+      else:
+        size(i32)
+
+def align(type):
+  match type
+    f32 = 4
+    f64 = 8
+    s8 = 1
+    u8 = 1
+    s16 = 2
+    u16 = 2
+    s32 = 4
+    u32 = 4
+    s64 = 8
+    u64 = 8
+    char = 4
+    string = 4
+    list $ty = 4
+    flags $fields = 4
+    record $fields =
+      a = 1
+      for field in $fields
+        a = max(a, align(field))
+      a
+    variant $cases =
+      a = align(variant_discriminant($cases))
+      for case in $cases
+        if case:
+          a = max(a, align(case))
+      a
+    handle => align(i32)
+    push-buffer | pull-buffer => align(i32)
+
+def variant_discriminant(cases):
+  match cases.len() {
+    0 => unreachable,
+    n if n <= 1<<8 => i8,
+    n if n <= 1<<16 => i16,
+    n if n <= 1<<32 => i32,
+    _ => i64,
+  }
+```
+
+With those definitions in mind the way to lift types from memory is then
+relatively straightforward:
+
+```
+def lift_from_memory(direction, src, ty, ptr):
+  match ty
+    f32 => yield(read_4_bytes(src, ptr) as f32),
+    f64 => yield(read_8_bytes(src, ptr) as f64),
+
+    s8 => yield(read_1_byte(src, ptr) as s8),
+    u8 => yield(read_1_byte(src, ptr) as u8),
+    s16 => yield(read_2_bytes(src, ptr) as s16),
+    u16 => yield(read_2_bytes(src, ptr) as u16),
+    s32 => yield(read_4_bytes(src, ptr) as s32),
+    u32 => yield(read_4_bytes(src, ptr) as u32),
+    s64 => yield(read_8_bytes(src, ptr) as s64),
+    u64 => yield(read_8_bytes(src, ptr) as u64),
+
+    # defer validation of the value to `lift`
+    char => yield from lift(direction, src, char, [read_4_bytes(src, ptr) as wasm_i32])
+
+    # defer the heavy-lifting to, well, `lift`
+    string |
+    list<$ty> => {
+      ptr = read_4_bytes(src, ptr) as wasm_i32
+      len = read_4_bytes(src, ptr + 4) as wasm_i32
+      yield from lift(direction, src, list<$ty>, [ptr, len])
+    }
+
+    flags<$fields> => {
+      values = []
+      for i in 0..num_packed_i32s($fields):
+        values.push(read_4_bytes(src, ptr + i * 4))
+      yield from lift(direction, src, flags<$fields>, values)
+    }
+
+    record<$fields> => {
+      yield(record_start<$fields>)
+      offset = 0
+      for $field in $fields:
+        offset = align_to(offset, align($field))
+        yield from lift_from_memory(direction, src, ptr + offset, $field)
+        offset += size(direction, $field)
+      yield(record_end<$fields>)
+    }
+
+    variant<$cases> => {
+      match val {
+        $case(payload) => {
```

Comment:
`Val` seems unbound? Either match $case is a typo and shouldn’t be here, or I am confused as to what this supposed to mean.